### PR TITLE
Update to 9.10 - release new versions

### PR DESCRIPTION
**Pull request by jasagredo**
# Description

Also some dev infra updates:

- Update actions in workflows
- Use `cabal haddock-project` instead of custom script
- Delete unused `cabal.project.local.haskell`
- Test GHC 9.10 on CI
- Update CI cabal-install to 3.12.1.0

# Checklist

- [X] Document and test your changes.
- [X] Update documentation such as `README.md` files.
- [X] Changes introduced by the PR should be recorded in the relevant
  `CHANGELOG.md` files.
- [X] Link issues that the PR resolves (if any).

### CODE REVIEW

File: diff-containers/diff-containers.cabal [hunk 1]
```diff
@@ -17,30 +17,53 @@ author:          Joris Dral
...
-library
+source-repository head
+  type:     git
+  location: https://github.com/input-output-hk/anti-diffs
+  subdir:   diff-containers
+
+flag asserts
+  description: Enable assertions
+  manual: False
+  default: False
```
Comment (jorisdral) [inline diff-containers/diff-containers.cabal:36]: I don't think we have to add this flag. We're not using this in our code

File: .github/workflows/documentation.yml [hunk 2]
```diff
@@ -67,11 +64,11 @@ jobs:
 
     - name: Build haddocks
       run: |
-        ./scripts/haddocks.sh
-        tar vzcf haddocks.tgz ./docs/haddocks
+        cabal haddock-project --internal --hackage --haddock-options="--title anti-diffs --package-name anti-diffs --gen-index --gen-contents --quickjump"
```
Comment (jorisdral) [inline .github/workflows/documentation.yml:67]: Have you checked locally that this produces roughly the same haddocks website?

File: .github/workflows/haskell.yml [hunk 3]
```diff
@@ -87,17 +88,18 @@ jobs:
       run: cabal build all
 
     - name: Run tests
-      run: cabal test -j1 --test-show-details=direct all
+      run: cabal test -j1 all
```
Comment (jorisdral) [inline .github/workflows/haskell.yml:91]: This is useful to keep

File: .github/workflows/haskell.yml [hunk 4]
```diff
@@ -87,17 +88,18 @@ jobs:
...
 
   # Check formatting for Haskell files
   stylish-haskell:
+    name: "stylish-haskell"
```
Comment (jorisdral) [inline .github/workflows/haskell.yml:95]: obsolete

File: .github/workflows/haskell.yml [hunk 5]
```diff
@@ -157,20 +159,21 @@ jobs:
         git diff --exit-code
 
   # Check formatting for cabal files
-  cabal-fmt:
+  cabal-gild:
```
Comment (jorisdral) [inline .github/workflows/haskell.yml:162]: I like `cabal-fmt`, I'm not sure it's necessary to switch

File: .github/workflows/haskell.yml (hunk 3 above)
Comment (jasagredo) [reply .github/workflows/haskell.yml:91]: It is the default in newer cabals.

File: .github/workflows/haskell.yml (hunk 5 above)
Comment (jorisdral) [reply .github/workflows/haskell.yml:162]: It's less reader friendly -- we're not changing these cabals that often that the diff matters

File: diff-containers/diff-containers.cabal (hunk 1 above)
Comment (jorisdral) [reply diff-containers/diff-containers.cabal:36]: My point is that the flag is `False`, never set to `True`, and we don't use `-DENABLE_ASSERTIONS` in our code

If we want to run our tests in CI with assertions enabled, then IMO we should pass `-fno-ignore-asserts` in CI. Custom flags on a library are also not great

Comment (jasagredo) [reply diff-containers/diff-containers.cabal:36]: Ah I see what you mean. True, lets remove it

File: .github/workflows/haskell.yml (hunk 5 above)
Comment (jasagredo) [reply .github/workflows/haskell.yml:162]: Ok. Reverting to cabal-fmt

File: diff-containers/diff-containers.cabal [hunk 6]
```diff
@@ -17,30 +17,44 @@ author:          Joris Dral
 maintainer:      operations@iohk.io, Joris Dral (joris@well-typed.com)
 category:        Data Structures
 build-type:      Simple
-tested-with:     GHC ==8.10 || ==9.2 || ==9.4 || ==9.6 || ==9.8
+tested-with:     GHC ==8.10 || ==9.6 || ==9.10
```
Comment (jorisdral) [inline diff-containers/diff-containers.cabal:20]: This should still include `9.2, `9.4` and `9.8`

File: fingertree-rm/fingertree-rm.cabal [hunk 7]
```diff
@@ -17,50 +17,59 @@ author:          Joris Dral
 maintainer:      operations@iohk.io, Joris Dral (joris@well-typed.com)
 category:        Data Structures
 build-type:      Simple
-tested-with:     GHC ==8.10 || ==9.2 || ==9.4 || ==9.6 || ==9.8
+tested-with:     GHC ==8.10 || ==9.6 || ==9.10
```
Comment (jorisdral) [inline fingertree-rm/fingertree-rm.cabal:20]: Idem

**Review by jorisdral — APPROVED** [review]
LGTM -- just two small things to change and then you can merge

**Pull request merged**